src/data/releases/release.service.ts

```diff
@@ -19,57 +19,377 @@
 
 import { Release, ReleaseState, ReleaseMongooseDocument, ReleaseFilesInput } from './release.model';
 import * as releaseModel from './release.model';
-import logger from '../../logger';
+import Logger from '../../logger';
+const logger = Logger('Release.DataService');
+
+// Values are the list of states that the release can be in to allow transition to the property name
+const ALLOWED_STATE_TRANSITIONS: { [key in ReleaseState]: ReleaseState[] } = {
+  [ReleaseState.CREATED]: [],
+  [ReleaseState.CALCULATING]: [
+    ReleaseState.CREATED,
+    ReleaseState.CALCULATED,
+    ReleaseState.BUILT,
+    ReleaseState.ERROR_BUILD,
+    ReleaseState.ERROR_CALCULATE,
+    ReleaseState.ERROR_PUBLISH,
+  ],
+  [ReleaseState.CALCULATED]: [ReleaseState.CALCULATING],
+  [ReleaseState.ERROR_CALCULATE]: [ReleaseState.CALCULATING],
+  [ReleaseState.BUILDING]: [
+    ReleaseState.CALCULATED,
+    ReleaseState.BUILT,
+    ReleaseState.ERROR_BUILD,
+    ReleaseState.ERROR_PUBLISH,
+  ],
+  [ReleaseState.BUILT]: [ReleaseState.BUILDING],
+  [ReleaseState.ERROR_BUILD]: [ReleaseState.BUILDING],
+  [ReleaseState.PUBLISHING]: [ReleaseState.BUILT, ReleaseState.ERROR_PUBLISH],
+  [ReleaseState.PUBLISHED]: [ReleaseState.PUBLISHING],
+  [ReleaseState.ERROR_PUBLISH]: [ReleaseState.PUBLISHING],
+};
+
+function isAllowedStateTransition(release: Release, newState: ReleaseState): boolean {
+  return ALLOWED_STATE_TRANSITIONS[newState].includes(release.state);
+}
```

Comment:
Convenience method to check if a given state can be transitioned into for the current release.